server/devdocs/bin/generate-components-usage-counts

```diff
@@ -0,0 +1,174 @@
+#!/usr/bin/env node
+
+/**
+ * This script generates a usage counts for the dependecies of a list of modules.
+ * It accepts a newline-delimited list of .js and|or .jsx files
+ * as its input, and writes the index to server/devdocs/usage-counts.js
+ */
+
+var async = require( 'async' ),
+	fs = require( 'fs' ),
+	fspath = require( 'path' ),
+	root = fspath.dirname( fspath.join( __dirname, '..', '..' ) ),
+
+	// Copyright (c) 2014-present, Facebook, Inc. See CREDITS.md#L1383
+	replacePatterns = {
+		BLOCK_COMMENT_RE: /(?:\/\*(.|\n)*?\*\/)|(\/\/.+(\n|$))/g,
+		LINE_COMMENT_RE: /\/\/.+(\n|$)/g,
+		IMPORT_RE: /(\bimport\s+?(?:.+\s+?from\s+?)?)(['"])([^'"]+)(\2)/g,
+		EXPORT_RE: /(\bexport\s+(?:[^'"]+\s+from\s+)??)(['"])([^'"]+)(\2)/g,
+		REQUIRE_RE: /(\brequire\s*?\(\s*?)(['"])([^'"]+)(\2\s*?\))/g
+	};
+
+function main() {
+	// extract list of files to index and remove leading ./'s
+	var fileList = process.
+		argv.
+		splice( 2, process.argv.length ).
+		map( function ( fileWithPath ) {
+			return fileWithPath.replace( /^\.\//, '' );
+		} );
+
+	if ( fileList.length === 0 ) {
+		process.stderr.write( 'You must pass a list of files to process (try "make server/devdocs/usage-counts.js"' );
+		process.exit( 1 );
+	}
+
+	getModulesWithDependencies( root, fileList )
+		.then(  function ( modulesWithDependencies ) {
+			var stats = generateStats( modulesWithDependencies );
+			writeStats( stats, 'server/devdocs/usage-counts.js' );
+			process.exit( 0 );
+		} )
+		.catch( function ( error ) {
+			console.error( "An error occurred while processing the files: \n\t", error );
+			process.exit( 1 );
+		} );
+}
+
+/**
+ * Calculate the modules dependencies.
+ * Produces an object that has the following structure:
+ * {
+ *  'A': [dep1, dep2, dep3, ...],
+ *  'B': [dep1, dep2, dep3, ...]
+ * }
+ * @param {string} root BasePath to resolve the files' path
+ * @param {Array} fileList A list of file names
+ * @returns {Promise}
+ */
+
+function getModulesWithDependencies( root, fileList ) {
+	return new Promise( function ( resolve, reject ) {
+		var filesContent = {},
+		dependencies = {};
+
+		async.each(
+			fileList,
+			function ( fileWithPath, next ) {
+				fs.readFile(
+					fspath.join( root, fileWithPath ),
+					{ encoding: 'utf8' },
+					function ( err, data ) {
+						if ( err ) {
+							return next( err );
+						}
+						// get the dependencies
+						filesContent[ fileWithPath ] = data;
+						next( null );
+					}
+				);
+			},
+			function ( err ) {
+				if ( err ) {
+					return reject( err );
+				}
+				Object.keys( filesContent ).reduce(function ( target, module ) {
```

Comment:
Why do we need to calculate dependencies after we have read all the files and not right after reading each one of them?